jiant/tasks/tasks.py

```diff
@@ -3731,3 +3732,171 @@ def get_metrics(self, reset=False):
+            return set(tokenized_eod).issubset(set(seg))
+
+        f = open(path, "r")
+        #  The dataset comes with one sentence per line, thus we split by
+        #  line here.
+        current_chunk = [_tokenize(self._tokenizer_name, next(f))]
+        current_length = len(current_chunk[0])
+        target_seq_length = self.get_target_seq_length()
+        while len(current_chunk) > 0:
+            segment = next(f)
+            segment = _tokenize(self._tokenizer_name, segment)
+            if is_end_document(segment) or current_length >= target_seq_length:
+                for_next_chunk = []
+                if current_length > target_seq_length:
+                    # Since the most current sentence added to the chunk exceeds the target
+                    # length, we save it for the next chunk (next example).
+                    for_next_chunk.append(current_chunk.pop())
+                if not is_end_document(segment):
+                    for_next_chunk.append(segment)
+                target_seq_length = self.get_target_seq_length()
+                if len(current_chunk) >= 2:
+                    # Make sure we have at least 2 sentences to distribute between the two
+                    # segments.
+                    a_end = random.randint(1, len(current_chunk) - 1)
+                    tokens_a = []
+                    for j in range(a_end):
+                        tokens_a.extend(current_chunk[j])
+                    tokens_b = []
+                    for j in range(a_end, len(current_chunk)):
+                        tokens_b.extend(current_chunk[j])
+                    in_order = random.random() < 0.5
+                    if in_order:
+                        yield (tokens_a, tokens_b, in_order)
+                    else:
+                        yield (tokens_b, tokens_a, in_order)
+                # if len(current_chunk) >=2, we will yield and reinitialize
+                # if len(current_chunk) ==1, we will not yeild, and reinitialize
+                if len(for_next_chunk) > 0 and not is_end_document(segment):
+                    # Make sure we only sample articles for each example that
+                    # belong to the same document.
+                    current_chunk = for_next_chunk
+                    current_length = sum([len(chunk) for chunk in for_next_chunk])
+                else:
+                    # We find the next sentence for the next example.
+                    try:  # Might run into StopIterationError
+                        current_chunk = [_tokenize(self._tokenizer_name, next(f))]
+                        current_length = len(current_chunk[0])
+                    except:
+                        print("Done loading data for SOP")
+                        current_chunk = []
+                        current_length = 0
+                        pass
+            else:
+                current_chunk.append(segment)
+                current_length += len(segment)
```

Comment:
I've been trying to understand this logic, and I'm struggling to understand the corner cases.

Based on the description in the docstring I tried to replicate the logic as I understand it. This is what I came up with:

```
        tokenizer = get_tokenizer(self._tokenizer_name)
        with open(path, "r") as f:
            document = []
            for line in f:
                line = line.strip()
                if "END OF ARTICLE" not in line:
                    tokenized_line = tokenizer.tokenize(line)
                    document.append(tokenized_line)
                else:
                    if len(document) < 2:
                        continue
                    for first_token_seq, second_token_seq in zip(document, document[1:]):
                        yield first_token_seq, second_token_seq, True
                        yield second_token_seq, first_token_seq, False
                    document = []
```

This doesn't implement the "90%" logic I asked about [above](https://github.com/nyu-mll/jiant/pull/1061/files#r409929289), but it passes the test provided in your PR.

Can you help me understand what I'm missing that's implemented in L3789-L3849?

Reply:
OK, since last night this has come into focus — this code is mostly ported [from TensorFlow](https://github.com/google-research/albert/blob/4135001fb1a84491bb86bebb2646523495d36f74/create_pretraining_data.py#L273-L350).

Instead of copying this over and rewriting components, please ~either import (probably not worth the overhead of making this a dependency), or~ directly copy the TensorFlow implementation into a utility file (and provide an attribution/license/modification notice for the copied code), then make the minimal set of changes to that encapsulated code to make compatible with your task (commenting your modifications).

Reply:
There's a few things incorrect. First, with 50% chance, you yield first_token_seq, second_token_seq, True and 50% chance you yield second_token_seq, first_token_seq, False. Second, you need to get, at each example, as many sentences as you can and still be under the target_seq_length limit (which 90% of the time is max_seq_len and 10% of the time is a random number between 2 and max_seq_len).

Reply:
Sorry, I still don't understand — I need to set up time with you to discuss. I'll ping you.